### PR TITLE
Show install missing packages button in open workspace dialog for plugins

### PR DESCRIPTION
**Pull request by casperlamboo**
Screenshot 2023-07-27 at 10 15 57

Screenshot 2023-07-27 at 10 16 50

Cura-10719

### CODE REVIEW

File: plugins/3MFWriter/ThreeMFWriter.py [hunk 1]
```diff
@@ -253,7 +257,64 @@ def _storeMetadataJson(metadata: Dict[str, List[Dict[str, str]]], archive: zipfi
         metadata_file = zipfile.ZipInfo(path)
         # We have to set the compress type of each file as well (it doesn't keep the type of the entire archive)
         metadata_file.compress_type = zipfile.ZIP_DEFLATED
-        archive.writestr(metadata_file, json.dumps(metadata, separators=(", ", ": "), indent=4, skipkeys=True, ensure_ascii=False))
+        archive.writestr(metadata_file,
+                         json.dumps(metadata, separators=(", ", ": "), indent=4, skipkeys=True, ensure_ascii=False))
+
+    @staticmethod
+    def _getPluginPackageMetadata() -> List[Dict[str, str]]:
```
Comment (nallath) [inline plugins/3MFWriter/ThreeMFWriter.py:264]: I think this function should have a few tests.

File: plugins/3MFWriter/ThreeMFWriter.py [hunk 2]
```diff
@@ -253,7 +257,64 @@ def _storeMetadataJson(metadata: Dict[str, List[Dict[str, str]]], archive: zipfi
...
+
+        plugin_ids = set()
+
+        def add_plugin_ids_in_stack(stack: CuraContainerStack):
+            for key in stack.getAllKeys():
+                value = str(stack.getProperty(key, "value"))
+                for plugin_id, _version, _value in backend_plugin_enum_value_regex.findall(value):
+                    plugin_ids.add(plugin_id)
+
+        # go through all stacks and find all the plugin id contained in the project
```
Comment (nallath) [inline plugins/3MFWriter/ThreeMFWriter.py:288]:
```suggestion
        # Go through all stacks and find all the plugin id contained in the project
```

File: plugins/3MFWriter/ThreeMFWriter.py [hunk 3]
```diff
@@ -253,7 +257,64 @@ def _storeMetadataJson(metadata: Dict[str, List[Dict[str, str]]], archive: zipfi
...
+
+        for container in global_stack.getContainers():
+            add_plugin_ids_in_stack(container)
+
+        for extruder_stack in global_stack.extruderList:
+            add_plugin_ids_in_stack(extruder_stack)
+
+            for container in extruder_stack.getContainers():
+                add_plugin_ids_in_stack(container)
+
```
Comment (nallath) [inline plugins/3MFWriter/ThreeMFWriter.py:300]:
```suggestion
        for stack in chain(global_stack, global_stack.extruderList):
                add_plugin_ids_in_stack(stack)

                for container in stack.getContainers():
                        add_plugin_ids_in_stack(container)
```

Less code duplication.

**Review by nallath — DISMISSED** [review]

File: plugins/3MFReader/WorkspaceDialog.py [hunk 4]
```diff
@@ -408,26 +408,27 @@ def waitForClose(self) -> None:
...
+            button_style=Message.ActionButtonStyle.LINK
         )
         result_message.addAction(
-                "install_materials",
-                name=i18n_catalog.i18nc("@action:button", "Install Materials"),
-                icon="",
-                description="Install missing materials from project file.",
-                button_align=Message.ActionButtonAlignment.ALIGN_RIGHT,
-                button_style=Message.ActionButtonStyle.DEFAULT
+            "install_materials",
```
[the remark this answers is not in the document]
Comment (casperlamboo) [reply plugins/3MFReader/WorkspaceDialog.py:426]: It's never displayed to the user right? But you're right; I'll change it
```suggestion
            "install_packages",
```

File: plugins/3MFWriter/ThreeMFWriter.py [hunk 5]
```diff
@@ -253,7 +257,64 @@ def _storeMetadataJson(metadata: Dict[str, List[Dict[str, str]]], archive: zipfi
...
+        def add_plugin_ids_in_stack(stack: CuraContainerStack):
+            for key in stack.getAllKeys():
+                value = str(stack.getProperty(key, "value"))
+                for plugin_id, _version, _value in backend_plugin_enum_value_regex.findall(value):
+                    plugin_ids.add(plugin_id)
+
+        # go through all stacks and find all the plugin id contained in the project
+        global_stack = Application.getInstance().getMachineManager().activeMachine
+        add_plugin_ids_in_stack(global_stack)
```
Comment (casperlamboo) [inline plugins/3MFWriter/ThreeMFWriter.py:290]:
```suggestion
        addPluginIdsInStack(global_stack)
```

File: plugins/3MFWriter/ThreeMFWriter.py [hunk 6]
```diff
@@ -253,7 +257,64 @@ def _storeMetadataJson(metadata: Dict[str, List[Dict[str, str]]], archive: zipfi
...
+
+        for container in global_stack.getContainers():
+            add_plugin_ids_in_stack(container)
```
Comment (casperlamboo) [inline plugins/3MFWriter/ThreeMFWriter.py:293]:
```suggestion
            addPluginIdsInStack(container)
```

File: plugins/3MFWriter/ThreeMFWriter.py [hunk 7]
```diff
@@ -253,7 +257,64 @@ def _storeMetadataJson(metadata: Dict[str, List[Dict[str, str]]], archive: zipfi
...
+
+        for extruder_stack in global_stack.extruderList:
+            add_plugin_ids_in_stack(extruder_stack)
```
Comment (casperlamboo) [inline plugins/3MFWriter/ThreeMFWriter.py:296]:
```suggestion
            addPluginIdsInStack(extruder_stack)
```

File: plugins/3MFWriter/ThreeMFWriter.py [hunk 8]
```diff
@@ -253,7 +257,64 @@ def _storeMetadataJson(metadata: Dict[str, List[Dict[str, str]]], archive: zipfi
...
+
+            for container in extruder_stack.getContainers():
+                add_plugin_ids_in_stack(container)
```
Comment (casperlamboo) [inline plugins/3MFWriter/ThreeMFWriter.py:299]:
```suggestion
                addPluginIdsInStack(container)
```

**Pull request merged**